Ballerina-Training/Ballerina-Forum-Application/back-end/bal-forum/main.bal

```diff
@@ -0,0 +1,222 @@
+                body: {
+                    error_message: "User already exists"
+                }
+            };
+            return userExists;
+        }
+
+        string id = uuid:createType1AsString();
+        _ = check forumDbClient->execute(`
+            INSERT INTO users VALUES (${id}, ${newUser.name}, ${newUser.email}, ${newUser.password}, "[]")
+        `);
+
+        _ = start sendEmailNotification(newUser);
+
+        Created userCreated = {
+            body: {
+                message: "User created successfully!"
+            }
+        };
+        return userCreated;
+    }
+
+    resource function get users/[string id]() returns User|NotFound|error {
+        UserDb|error userDb = forumDbClient->queryRow(`
+            SELECT name, email, id, subscribtions FROM users WHERE id = ${id}
+        `);
+        if userDb is UserDb {
+            User user = check transformUserFromDatabase(userDb);
+            return user;
+        }
+
+        NotFound userNotFound = {
+            body: {
+                error_message: "User not found"
+            }
+        };
+        return userNotFound;
+    }
+
+    resource function post login(UserLogin userLogin) returns LoginOk|Unauthorized {
+        User|error user = forumDbClient->queryRow(`
+            SELECT id, name, email FROM users WHERE name = ${userLogin.name} && password = ${userLogin.password}
+        `);
+        if user is User {
+            LoginOk loginOk = {
+                body: {
+                    message: "Login successfully",
+                    user: user
+                }
+            };
+            return loginOk;
+        }
+
+        Unauthorized unauthorized = {
+            body: {
+                error_message: "Invalid credentials"
+            }
+        };
+        return unauthorized;
+    }
+
+    resource function post users/[string id]/posts(NewPost newPost) returns Created|Forbidden|error {
+        Sentiment sentiment = check sentimentClient->/api/sentiment.post({
+            text: newPost.description
+        });
+        if sentiment.label == "neg" {
+            Forbidden forbidden = {
+                body: {
+                    error_message: "Post rejected due to negative sentiment"
+                }
+            };
+            return forbidden;
+        }
+
+        UserDb userDb = check forumDbClient->queryRow(`
+            SELECT name, email, id, subscribtions FROM users WHERE id = ${id}
+        `);
+
+        User user = check transformUserFromDatabase(userDb);
+        Post post = check createPostFromNewPost(newPost, user.name);
+
+        _ = check forumDbClient->execute(`
+            INSERT INTO posts VALUES (${post.id}, ${post.title}, ${post.description}, ${post.username}, ${post.likes.toJsonString()}, ${post.comments.toJsonString()}, ${post.postedAt})
+        `);
+
+        if user.subscribtions.indexOf(post.id) is () {
+            user.subscribtions.push(post.id);
+            _ = check forumDbClient->execute(`
```

Comment:
Whenever we create new posts, we update the users subscriptions as well. So these two DB operations have to be within a transaction right?

Reply:
Ack

Reply:
Fixed via [65b9c57](https://github.com/ballerina-guides/training-content/pull/12/commits/65b9c5708ef9ba4bd4e065a5b33c94efe5fc1e65)